apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -778,91 +813,162 @@ handle_query_result(Id, Result, HasBeenSent) ->
                 fun() ->
                     ?SLOG(error, #{id => Id, msg => unrecoverable_error, reason => Reason}),
-                    inc_sent_failed(Id, HasBeenSent),
                     ok
                 end,
-            {ack, PostFn};
+            Counters =
+                case HasBeenSent of
+                    true -> #{retried_failed => 1};
+                    false -> #{failed => 1}
+                end,
+            {ack, PostFn, Counters};
         false ->
             PostFn = fun() ->
                 ?SLOG(error, #{id => Id, msg => async_send_error, reason => Reason}),
                 ok
             end,
-            {nack, PostFn}
+            {nack, PostFn, #{}}
     end;
 handle_query_async_result_pure(_Id, {ok, Pid}, _HasBeenSent) when is_pid(Pid) ->
-    {ack, fun() -> ok end};
+    {ack, fun() -> ok end, #{}};
 handle_query_async_result_pure(_Id, ok, _HasBeenSent) ->
-    {ack, fun() -> ok end}.
+    {ack, fun() -> ok end, #{}}.
+
+-spec aggregate_counters(data(), counters()) -> data().
+aggregate_counters(Data = #{counters := OldCounters}, DeltaCounters) ->
+    Counters = merge_counters(OldCounters, DeltaCounters),
+    Data#{counters := Counters}.
+
+-spec merge_counters(counters(), counters()) -> counters().
+merge_counters(OldCounters, DeltaCounters) ->
+    maps:fold(
+        fun(Metric, Val, Acc) ->
+            maps:update_with(Metric, fun(X) -> X + Val end, Val, Acc)
+        end,
+        OldCounters,
+        DeltaCounters
+    ).
+
+-spec flush_metrics(data()) -> data().
+flush_metrics(Data = #{id := Id, counters := Counters}) ->
+    bump_counters(Id, Counters),
+    set_gauges(Data),
+    ensure_metrics_flush_timer(Data#{counters := #{}}).
+
+-spec ensure_metrics_flush_timer(data()) -> data().
+ensure_metrics_flush_timer(Data = #{metrics_tref := undefined, metrics_flush_interval := T}) ->
+    Ref = make_ref(),
+    TRef = erlang:send_after(T, self(), {flush_metrics, Ref}),
+    Data#{metrics_tref := {TRef, Ref}}.
+
+-spec bump_counters(id(), counters()) -> ok.
+bump_counters(Id, Counters) ->
+    maps:foreach(
+        fun
+            (dropped_expired, Val) ->
+                emqx_resource_metrics:dropped_expired_inc(Id, Val);
+            (dropped_queue_full, Val) ->
+                emqx_resource_metrics:dropped_queue_full_inc(Id, Val);
+            (failed, Val) ->
+                emqx_resource_metrics:failed_inc(Id, Val);
+            (retried_failed, Val) ->
+                emqx_resource_metrics:retried_failed_inc(Id, Val);
+            (success, Val) ->
+                emqx_resource_metrics:success_inc(Id, Val);
+            (retried_success, Val) ->
+                emqx_resource_metrics:retried_success_inc(Id, Val);
+            (dropped_resource_not_found, Val) ->
+                emqx_resource_metrics:dropped_resource_not_found_inc(Id, Val);
+            (dropped_resource_stopped, Val) ->
+                emqx_resource_metrics:dropped_resource_stopped_inc(Id, Val)
+        end,
```

Comment:
write it as a static fun should be easier to read (and more performant).

Reply:
I think it'll still require a closure, since we're capturing `Id` from the context.  :thinking:

Reply:
Alternatively, we could use `maps:iterator` and recurse over that.  I think I'll do that to avoid the closure.